protocol/directory.go

```diff
@@ -0,0 +1,147 @@
+	d.policies = merkletree.NewPolicies(epDeadline, vrfKey)
+}
+
+func (d *ConiksDirectory) LatestSTR() *merkletree.SignedTreeRoot {
+	return d.pad.LatestSTR()
+}
+
+// HandleOps validates the request message and then pass it to
+// appropriate operation handler according to the request type.
+func (d *ConiksDirectory) HandleOps(req *Request) (Response, ErrorCode) {
+	switch req.Type {
+	case RegistrationType:
+		if msg, ok := req.Request.(*RegistrationRequest); ok {
+			if len(msg.Username) > 0 && len(msg.Key) > 0 {
+				return d.Register(msg)
+			}
+		}
+	case KeyLookupType:
+		if msg, ok := req.Request.(*KeyLookupRequest); ok {
+			if len(msg.Username) > 0 {
+				return d.KeyLookup(msg)
+			}
+		}
+	case ConsistencyCheckType:
+		if msg, ok := req.Request.(*KeyLookupRequest); ok {
+			// we don't expect to have a consistency check at epoch 0
+			if len(msg.Username) > 0 && msg.Epoch > 0 {
+				return d.ConsistencyCheck(msg)
+			}
+		}
+	}
+	return nil, ErrorMalformedClientMessage
+}
+
+func (d *ConiksDirectory) Register(req *RegistrationRequest) (
+	*RegistrationProof, ErrorCode) {
+	// check whether the name already exists
+	// in the directory before we register
+	ap, err := d.pad.Lookup(req.Username)
+	if err != nil {
+		return nil, ErrorDirectory
+	}
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		return NewRegistrationProof(ap, d.LatestSTR(), nil), ErrorNameExisted
+	}
+	if d.useTBs {
+		// also check the temporary bindings array
+		// currently the server allows only one registration/key change per epoch
+		if tb := d.tbs[req.Username]; tb != nil {
+			return NewRegistrationProof(ap, d.LatestSTR(), tb), ErrorNameExisted
+		}
+
+		// insert new data to the directory on-the-fly
+		tb, err := d.pad.TB(req.Username, req.Key)
+		if err != nil {
+			return nil, ErrorDirectory
+		}
+		d.tbs[req.Username] = tb
+		return NewRegistrationProof(ap, d.LatestSTR(), tb), Success
+	} else {
+		if err = d.pad.Set(req.Username, req.Key); err != nil {
+			return nil, ErrorDirectory
+		}
+		return NewRegistrationProof(ap, d.LatestSTR(), nil), Success
+	}
+}
+
+func (d *ConiksDirectory) KeyLookup(req *KeyLookupRequest) (
+	*KeyLookupProof, ErrorCode) {
+	ap, err := d.pad.Lookup(req.Username)
+	if err != nil {
+		return nil, ErrorDirectory
+	}
+
+	if !bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		// if not found in the tree, do lookup in tb array
+		if d.useTBs {
+			if tb := d.tbs[req.Username]; tb != nil {
+				return NewKeyLookupProof(ap, d.LatestSTR(), tb), Success
+			}
+		}
+		return NewKeyLookupProof(ap, d.LatestSTR(), nil), ErrorNameNotFound
+	}
+	return NewKeyLookupProof(ap, d.LatestSTR(), nil), Success
+}
+
+// ConsistencyCheck uses the same request struct
+// as KeyLookup but with a different `Request`'s `Type`.
+// The caller must check for Epoch field when unmarshalling the request
+// before passing it into this method.
+func (d *ConiksDirectory) ConsistencyCheck(req *KeyLookupRequest) (
```

Comment:
> I don't know how useful this information would be.

That's why I said above,

> That's a type of consistency, though it can easily be defeated by targeted responses.

I guess my point was that it shouldn't be a reason for a different API.

> Not yet in ctypes-otr, though @c633 and I discussed that we'll need to do this

Great!